### PR TITLE
Test minimisation

### PR DESCRIPTION
**Pull request by cristina-david**

### CODE REVIEW

**Comment by kroening** [comment]
Can you use this as an opportunity to practice squashing commits?

File: src/cbmc/bmc_cover.cpp [hunk 1]
```diff
@@ -120,6 +123,88 @@ class bmc_covert:
   goal_mapt goal_map;
   typedef std::vector<testt> testst;
   testst tests;
+
```
Comment (peterschrammel) [inline src/cbmc/bmc_cover.cpp]: guard debugging code by #ifdef DEBUG or remove

File: src/cbmc/bmc_cover.cpp [hunk 2]
```diff
@@ -120,6 +123,88 @@ class bmc_covert:
...
+      {
+        if(it.covered_goals.empty()) break; // break if remaining test cases will not contribute anyway
+        goals_diff(it.covered_goals,tc.covered_goals); // remove goals that are covered by the picked one
+      }    
+    }
+
+    status() << "Reduced from " << tsi.size() << " to " 
+	     << tso.size() << " test cases " <<messaget::eom;
+
+    //std::cout << "After minimisation:" << std::endl;
```
Comment (peterschrammel) [inline src/cbmc/bmc_cover.cpp]: guard debug code by #ifdef DEBUG or remove

File: src/cbmc/bmc_cover.cpp [hunk 3]
```diff
@@ -303,12 +392,21 @@ bool bmc_covert::operator()()
              << (sat_stop-sat_start) << "s" << eom;
   }
   
+
+  // should we minimise testsuite?
+  if(!bmc.options.get_bool_option("disable-testsuite-minimisation"))
+  {
+    testst tests_min;
+    minimise_testsuite(tests, tests_min, goal_map.size());
```
Comment (peterschrammel) [inline src/cbmc/bmc_cover.cpp]: I suggest to modify in-place to get rid of the second parameter of minimse_testsuite and copying

File: src/cbmc/bmc_cover.cpp [hunk 4]
```diff
@@ -120,6 +123,88 @@ class bmc_covert:
...
+      if(tc.covered_goals.empty()) break; // break if it does not contribute (and so do the remaining ones)
+      tss.erase(tss.begin()); // remove test case on top
+      tso.push_back(tc); // add it to the output suite
+      no_covered+=tc.covered_goals.size();
+      if(no_covered==no_goals)
+        break; // we've covered all the goals
+      for(auto &it : tss)
+      {
+        if(it.covered_goals.empty()) break; // break if remaining test cases will not contribute anyway
+        goals_diff(it.covered_goals,tc.covered_goals); // remove goals that are covered by the picked one
```
Comment (peterschrammel) [inline src/cbmc/bmc_cover.cpp]: whitespace after ','

File: src/cbmc/bmc_cover.cpp [hunk 5]
```diff
@@ -120,6 +123,88 @@ class bmc_covert:
...
+
+// greedy n^2 algorithm by descending ordering
+  void minimise_testsuite(testst& tsi, testst& tso, const unsigned no_goals) 
+  {
+    testst tss;
+    unsigned no_covered=0;
+    std::vector<irep_idt> goals;
+    goals.clear();
+
+    //print_testsuite(tsi);
```
Comment (peterschrammel) [inline src/cbmc/bmc_cover.cpp]: guard debug code by #ifdef DEBUG or remove

File: src/cbmc/bmc_cover.cpp [hunk 6]
```diff
@@ -120,6 +123,88 @@ class bmc_covert:
...
+    copy_testsuite(tsi,tss);
```
Comment (peterschrammel) [inline src/cbmc/bmc_cover.cpp]: an assignment should copy it

File: src/cbmc/bmc_cover.cpp [hunk 7]
```diff
@@ -120,6 +123,88 @@ class bmc_covert:
...
+    to.clear();
+    for(testst::const_iterator it=from.begin(); it!=from.end(); it++) 
+    {
+      to.push_back(*it);
+    }
+  }
+
+// compute covered goals set difference
+  static void goals_diff(std::set<irep_idt> &goals1, 
+			 std::set<irep_idt> &goals2)
```
Comment (peterschrammel) [inline src/cbmc/bmc_cover.cpp]: goals2 can be made const

File: src/cbmc/bmc_cover.cpp [hunk 8]
```diff
@@ -120,6 +123,88 @@ class bmc_covert:
...
+      std::cout << "Goals in test " << i << " : " << std::endl;
+      for (auto &goal : test.covered_goals)
+        std::cout << id2string(goal) << ";";
+
+      std::cout << std::endl;
+      i++;
+    }
+  }
+  
+  static void copy_testsuite(const testst &from, testst &to)
```
Comment (peterschrammel) [inline src/cbmc/bmc_cover.cpp]: can be removed

File: src/cbmc/bmc_cover.cpp [hunk 9]
```diff
@@ -120,6 +123,88 @@ class bmc_covert:
...
+      if(no_covered==no_goals)
+        break; // we've covered all the goals
+      for(auto &it : tss)
+      {
+        if(it.covered_goals.empty()) break; // break if remaining test cases will not contribute anyway
+        goals_diff(it.covered_goals,tc.covered_goals); // remove goals that are covered by the picked one
+      }    
+    }
+
+    status() << "Reduced from " << tsi.size() << " to "
```
Comment (peterschrammel) [inline src/cbmc/bmc_cover.cpp]: just remember size of tsi instead of copying it

**Comment by cristina-david** [comment]
@peterschrammel : I've addressed your comments in the new, squashed, version of the PR.

**Comment by cristina-david** [comment]
@kroening : I've squashed all the commits together.

**Comment by cristina-david** [comment]
I'm closing this here as it needs to be moved to cbmc-testgen.

**Pull request closed without merging**